### PR TITLE
Allow Multiple Co-Variates

### PR DESCRIPTION
**Pull request by swashcap**
* **Problem:** We’re locked to a single co-variate, the `isControl` tag.
* **Solution:** Re-work local pre-processing. _(Bonus: fix faulty `y` and `X` assignment in `preprocess`.)_
* **Testing:**
  1. `git clone git@github.com:MRN-Code/decentralized-laplacian-ridge-regression.git`
  2. Check out the _feature/multiple-covariates_ branch
  3. Run `npm test` to ensure tests pass!
  4. _Bonus:_ Run `npm run simulate` to use _coinstac-simulator_ and mock data against the computation.
* **Related:** MRN-Code/coinstac#45

### CODE REVIEW

File: src/local.js [hunk 1]
```diff
@@ -1,18 +1,61 @@
...
+ * @returns {Array}
+ */
+function addBias(arr) {
+  if (!Array.isArray(arr) || !arr.length) {
+    throw new Error(`Expected ${arr} to be an array with length`);
+  } else if (!arr.every(Array.isArray)) {
+    throw new Error(`Expected every item of ${arr} to be an array`);
+  }
+
+  return arr.map(row => row.concat(1));
```
Comment (swashcap) [inline src/local.js]: This is _literally_ all `addBias` does, as far as I understand. Old test passes. No need to mutate arrays as matrices, just add the damn number.

File: src/local.js [hunk 2]
```diff
@@ -1,18 +1,61 @@
...
+/**
+ * Get normalized co-variate values.
+ *
+ * @todo This ignores non-numeric and non-boolean tag values. Determine
+ * an effective way to map strings and non-primitives to numbers.
+ *
+ * @param {Object} tags Hash of tag names to tag values
+ * @returns {Array}
+ */
+function getNormalizedTags(tags) {
```
Comment (swashcap) [inline src/local.js]: We should get our hands on some actual metadata to see what we’re dealing with. Objective and gradient calculation only accept numeric input; we’ll need to support string-to-number mapping somehow.

File: src/local.js [hunk 3]
```diff
@@ -33,37 +76,73 @@ module.exports = {
...
 
-    const userData = opts.userData;
-    // do some file parsing. here, we're actually using pre-processed data...
-    // so just add some bias and bail!
-    const result = {};
-    const X = userData.files.map((file) => {
-      const surf = new FreeSurfer({
-        string: fs.readFileSync(file.filename).toString(),
-      });
+    return Promise.all(
```
Comment (swashcap) [inline src/local.js]: I promise this will confuse non-JavaScript developers. Oh, well. _Async FTW._

**Comment by swashcap** [comment]
@rssk Don’t merge yet!

**Comment by swashcap** [comment]
In talking to Sergey, we discovered that the algorithm itself may be implemented improperly. The gradient and objective should decrease over iterations.

File: src/local.js (hunk 2 above)
Comment (rssk) [inline src/local.js]: Totally agreed

**Pull request merged**